src/abstractdataframe/reshape.jl

```diff
@@ -425,22 +464,33 @@ function _unstack(df::AbstractDataFrame, rowkeys::AbstractVector{Int},
         throw(ArgumentError("Missing value in variable :$(_names(df)[colkey]). " *
                             "Pass `allowmissing=true` to skip missings."))
     end
-    unstacked_val = [fill!(similar(valuecol,
-                                   promote_type(eltype(valuecol), typeof(fill)),
-                                   Nrow),
-                           fill) for _ in 1:Ncol]
-
-    mask_filled = falses(Nrow, Ncol)
-
     @assert length(rowref) == length(colref) == length(valuecol)
-    for (k, (row_id, col_id, val)) in enumerate(zip(rowref, colref, valuecol))
-        if !allowduplicates && mask_filled[row_id, col_id]
-            throw(ArgumentError("Duplicate entries in unstack at row $k for key "*
-                                "$(tuple((df[k, s] for s in rowkeys)...)) and variable $(colref_map[col_id]). " *
-                                "Pass allowduplicates=true to allow them."))
+
+    if allowduplicates isa Bool
+        unstacked_bool = [fill!(similar(valuecol,
+                                    promote_type(eltype(valuecol), typeof(fill)),
+                                    Nrow),
+                            fill) for _ in 1:Ncol]
+
+        mask_filled = falses(Nrow, Ncol)
+
+        for (k, (row_id, col_id, val)) in enumerate(zip(rowref, colref, valuecol))
+            if !allowduplicates && mask_filled[row_id, col_id]
+                throw(ArgumentError("Duplicate entries in unstack at row $k for key "*
+                                    "$(tuple((df[k, s] for s in rowkeys)...)) and variable $(colref_map[col_id]). " *
+                                    "Pass allowduplicates=true to allow them."))
+            end
+            unstacked_bool[col_id][row_id] = val
+            mask_filled[row_id, col_id] = true
         end
-        unstacked_val[col_id][row_id] = val
-        mask_filled[row_id, col_id] = true
+        unstacked_val = unstacked_bool
+    else
+        @assert allowduplicates isa Function
+        unstacked_function = [[Vector{eltype(valuecol)}() for _ in 1:Nrow] for _ in 1:Ncol]
+        for (k, (row_id, col_id, val)) in enumerate(zip(rowref, colref, valuecol))
+            push!(unstacked_function[col_id][row_id], val)
+        end
```

Comment:
This is quite inefficient, but I guess we don't have a better solution in general? For simple cases like `sum` (or more generally a `Reduce(op)` object that would be exported) we could perform the reduction on the fly, but that can be left for later.

Reply:
Yes. We have two choices:
* either special case standard reductions in `unstack`
* run something like (I use the name `op` to call `allowduplicates` function); this is a rough description of what would have to be done:
```
unstack(combine(groupby(df, [rowkeys; colkeys], value => op => value), rowkeys, colkeys, value)
````
(which will re-use the optimizations we have in `combine` but then we need decide that we only want to allow scalars to be returned by `op`; if we went this way and then `fill` would matter as non-existent combinations would not get passed to `op`)

What do you think is better (if it is not clear what I mean here please comment)? I think it is better to pass empty vector to `op` always (and add optimizations later).

Reply:
Difficult choice. I'm not sure it matters a lot though since both options allow doing the same thing relatively conveniently. Also, even if we base the implementation on `combine` under the hood, we could call the provided function on an empty vector for each non-existent combination, right?

Would using `combine(groupby(...))` be an efficient implementation in general? Or are there advantages to using a separate implementation for `unstack`?

FWIW, tidyverse's `pivot_wider` seems to use the fill value even when a function is passed:
```R
> df <- data.frame(id=c("1", "1", "1", "2"),
                   variable=c("Var1", "Var1", "Var2", "Var1"),
                   value=c(1, 10, 2, 3))

> pivot_wider(df, names_from=variable, values_from=value)
# A tibble: 2 × 3
  id    Var1      Var2     
  <chr> <list>    <list>   
1 1     <dbl [2]> <dbl [1]>
2 2     <dbl [1]> <NULL>   
Message d'avis :
Values from `value` are not uniquely identified; output will contain list-cols.
* Use `values_fn = list` to suppress this warning.
* Use `values_fn = {summary_fun}` to summarise duplicates.
* Use the following dplyr code to identify duplicates.
  {data} %>%
    dplyr::group_by(id, variable) %>%
    dplyr::summarise(n = dplyr::n(), .groups = "drop") %>%
    dplyr::filter(n > 1L) 

> pivot_wider(df, names_from=variable, values_from=value, values_fn=sum, values_fill=-10)
# A tibble: 2 × 3
  id     Var1  Var2
  <chr> <dbl> <dbl>
1 1        11     2
2 2         3   -10
```

But when storing a list of values in cells, the fill value doesn't seem to be used, though if the type doesn't match an error is thrown (apparently when computing the spec rather than when doing actual work):
```R
# Same as above but without warning
> pivot_wider(df, names_from=variable, values_from=value, values_fn=list)
# A tibble: 2 × 3
  id    Var1      Var2     
  <chr> <list>    <list>   
1 1     <dbl [2]> <dbl [1]>
2 2     <dbl [1]> <NULL>   

> pivot_wider(df, names_from=variable, values_from=value, values_fn=list, values_fill=list(1, 2, 3))
# A tibble: 2 × 3
  id    Var1      Var2     
  <chr> <list>    <list>   
1 1     <dbl [2]> <dbl [1]>
2 2     <dbl [1]> <NULL>   

> pivot_wider(df, names_from=variable, values_from=value, values_fn=list, values_fill=1)
Erreur : Can't convert <double> to <list>.
Run `rlang::last_error()` to see where the error occurred.

> rlang::last_error()
<error/vctrs_error_incompatible_type>
Can't convert <double> to <list>.
Backtrace:
  1. tidyr::pivot_wider(...)
  2. tidyr:::pivot_wider.data.frame(...)
  3. tidyr::pivot_wider_spec(...)
  4. vctrs::vec_cast(fill, value)
  6. vctrs::vec_default_cast(...)
  7. vctrs::stop_incompatible_cast(...)
  8. vctrs::stop_incompatible_type(...)
  9. vctrs:::stop_incompatible(...)
 10. vctrs:::stop_vctrs(...)
Run `rlang::last_trace()` to see the full context.
```

Reply:
> Your example from R shows that it is inconsistent unfortunately - right?

Not really. AFAICT the fill value is ignored when a function is passed, which is a bit confusing but consistent. The fact that an error is thrown when the fill value has a type that doesn't match despite its being unused is also confusing but doesn't have any practical consequences.

> I would base this decision based on what we find more useful. For example if user asked for sum as aggregator on integer values. Do we feel that for missing combination unstack should produce missing or 0 in this case by default?

Right, this is a good example. On the one hand, it would probably be clearer to give `missing` by default so that people notice that some combinations don't exist. On the other, `0` is a very reasonable value for empty combinations (which is why `sum` returns it). But some functions like `median` throw an error for empty inputs so people would have to do `allowduplicates=x -> isempty(x) ? NaN : median(x)` instead of just `fill=NaN, allowduplicates=median`. That's certainly not the end of the world.

If we want to be able to change the behavior in the future, a safe approach would be to use a special internal sentinel type for `fill` so that we can distinguish omitting the argument from passing `fill=missing`, and throw an error only when `missing` is passed. Not sure whether it's worth it.

Reply:
There is no rush with merging this PR. Therefore I would like to discuss the behavior we want and implement what we find reasonable.

Given your arguments I think that:
* `fill` should be always respected. missing combinations should be filled always with `fill`. The reason is that it is easy enough to pass custom `fill` if someone desires so.
* the additional benefit of respecting `fill` is that we can use `unstack` on the result of `combine` on `groupby` (as this is the fast option). The reason is that we strongly prefer `allowduplicates=sum` over e.g. `allowduplicates=x-> isempty(x) ? missing : sum(x)` because passing `sum` will trigger fast path in `combine` (and passing anonymous function will use slow path instead). This means that common operations like `length`, `sum`, `mean` etc. will be fast.

I change the PR to follow this consideration.